openfl-workspace/tf_3dunet_brats/src/dataloader.py

```diff
@@ -0,0 +1,284 @@
+
+            img_temp = np.rot90(img_temp)
+
+            # Normalize
+            img_temp = self.z_normalize_img(img_temp)
+
+            img[..., channel] = img_temp
+
+        msk = np.rot90(np.array(nib.load(mskfile).dataobj))
+        msk = np.expand_dims(msk, -1)
+
+        # labels: {
+        #      0: background,
+        #      1: edema,
+        #      2: non-enhancing tumor,
+        #      3: enhancing tumour}
+
+        # Combine all masks but background
+        if self.num_classes == 1:
+            msk[msk > 0] = 1.0
+        else:
+            msk_temp = np.zeros(list(msk.shape) + [self.num_classes])
+            for channel in range(self.num_classes):
+                msk_temp[msk == channel, channel] = 1.0
+            msk = msk_temp
+
+        # Crop
+        img, msk = self.crop(img, msk, randomize)
+
+        # Randomly rotate
+        if randomize:
+            img, msk = self.augment_data(img, msk)
+
+        return img, msk
+
+    def get_input_shape(self):
+        """Get the shape of the input."""
+        return self.crop_dim
+
+    def plot_images(self, ds, slice_num=90):
+        """Plot images from dataset."""
+        import matplotlib.pyplot as plt
+
+        plt.figure(figsize=(20, 20))
+
+        num_cols = 2
+
+        msk_channel = 0
+        img_channel = 0
+
+        for img, msk in ds.take(1):
+            bs = img.shape[0]
+
+            for idx in range(bs):
+                plt.subplot(bs, num_cols, idx * num_cols + 1)
+                plt.imshow(img[idx, :, :, slice_num, img_channel], cmap='bone')
+                plt.title('MRI', fontsize=18)
+                plt.subplot(bs, num_cols, idx * num_cols + 2)
+                plt.imshow(msk[idx, :, :, slice_num, msk_channel], cmap='bone')
+                plt.title('Tumor', fontsize=18)
+
+        plt.show()
+
+        print(f'Mean pixel value of image = {np.mean(img[0, :, :, :, 0])}')
+
+    def display_train_images(self, slice_num=90):
+        """Plot some training images."""
+        self.plot_images(self.ds_train, slice_num)
+
+    def display_validation_images(self, slice_num=90):
+        """Plot some validation images."""
+        self.plot_images(self.ds_val, slice_num)
+
+    def display_test_images(self, slice_num=90):
+        """Plot some test images."""
+        self.plot_images(self.ds_test, slice_num)
+
+    def get_train(self):
+        """Return train dataset."""
+        return self.ds_train
+
+    def get_test(self):
+        """Return test dataset."""
+        return self.ds_test
+
+    def get_validate(self):
+        """Return validation dataset."""
+        return self.ds_val
+
+    def get_dataset(self):
+        """Create a TensorFlow data loader."""
+        self.num_train = int(self.numfiles * self.train_test_split)
+        numvaltest = self.numfiles - self.num_train
+
+        ds = tf.data.Dataset.range(self.numfiles).shuffle(
+            self.numfiles, self.random_seed)  # Shuffle the dataset
```

Comment:
Here I receive the following error:
```
File "/home/itrushkin/repos/openfl/fed_work12345alpha81671/src/dataloader.py", line 246, in get_dataset
    ds = tf.data.Dataset.range(self.numfiles).shuffle(
  File "/home/itrushkin/.virtualenvs/openfl/lib/python3.8/site-packages/tensorflow/python/data/ops/dataset_ops.py", line 1430, in shuffle
    return ShuffleDataset(self, buffer_size, seed, reshuffle_each_iteration)
  File "/home/itrushkin/.virtualenvs/openfl/lib/python3.8/site-packages/tensorflow/python/data/ops/dataset_ops.py", line 4022, in __init__
    variant_tensor = gen_dataset_ops.shuffle_dataset_v3(
  File "/home/itrushkin/.virtualenvs/openfl/lib/python3.8/site-packages/tensorflow/python/ops/gen_dataset_ops.py", line 6636, in shuffle_dataset_v3
    _ops.raise_from_not_ok_status(e, name)
  File "/home/itrushkin/.virtualenvs/openfl/lib/python3.8/site-packages/tensorflow/python/framework/ops.py", line 6897, in raise_from_not_ok_status
    six.raise_from(core._status_to_exception(e.code, message), None)
  File "<string>", line 3, in raise_from
tensorflow.python.framework.errors_impl.InvalidArgumentError: buffer_size must be greater than zero. [Op:ShuffleDatasetV3]
```
`self.numfiles` property is equal to zero. I suppose the glob `*/*_seg.nii.gz` doesn't match any file inside `data_path` directory. After executing `split_directory.sh`, folder structure of `${NEW_PATH}` is extended with additional directory level:
```
tree ${NEW_PATH}
${NEW_PATH}
├── split_0
│   ├── BraTS20_Training_001
│   │   └── BraTS20_Training_001        <---     single subfolder inside each folder
│   │       ├── BraTS20_Training_001_flair.nii
│   │       ├── BraTS20_Training_001_seg.nii
│   │       ├── BraTS20_Training_001_t1ce.nii
│   │       ├── BraTS20_Training_001_t1.nii
│   │       └── BraTS20_Training_001_t2.nii
│   ├── BraTS20_Training_003
│   │   └── BraTS20_Training_003        <---
│   │       ├── BraTS20_Training_003_flair.nii
│   │       ├── BraTS20_Training_003_seg.nii
│   │       ├── BraTS20_Training_003_t1ce.nii
│   │       ├── BraTS20_Training_003_t1.nii
│   │       └── BraTS20_Training_003_t2.nii
│   ├── BraTS20_Training_006
│   │   └── BraTS20_Training_006        <---
│   │       ├── BraTS20_Training_006_flair.nii
│   │       ├── BraTS20_Training_006_seg.nii
...
```

Reply:
I put in an exception if it can't find any BraTS files in the directory.

Reply:
I followed the steps in README.md and the error shouldn't take place. Probably there is something wrong in `split_directory.sh`. It shouldn't create multiple levels of folders. The folder structure should be:
```
tree ${NEW_PATH}
${NEW_PATH}
├── split_0
│   ├── BraTS20_Training_001
│   │   ├── BraTS20_Training_001_flair.nii
│   │   ├── BraTS20_Training_001_seg.nii
│   │   ├── BraTS20_Training_001_t1ce.nii
│   │   ├── BraTS20_Training_001_t1.nii
│   │   └── BraTS20_Training_001_t2.nii
│   ├── BraTS20_Training_003
│   │   ├── BraTS20_Training_003_flair.nii
...
```